File: vol1/ch6/LeetCode637.py
from collections import deque
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        ans=[]
        qu=deque()      #定义一个队列qu
        qu.append(root)              #根结点(层次为1)进队
        while qu:                    #队不空循环
            n=len(qu)
            cursum=0.0
            for i in range(0,n):
                p=qu.popleft()           #出队一个结点
                cursum+=p.val                 #当前层的结点值累计到cursum
                if p.left!=None:                      #有左孩子时将其进队
                    qu.append(p.left)
                if p.right!=None:                     #有右孩子时将其进队
                    qu.append(p.right)
            ans.append(cursum/n)
        return ans

File: vol1/ch6/LeetCode637.cpp
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> ans;
        queue<TreeNode*> qu;
        qu.push(root);                              //根结点进队
        while (!qu.empty()) {                           //队不空循环
            int n=qu.size();
            double cursum=0.0;
            for(int i=0;i<n;i++) {
                TreeNode*p=qu.front(); qu.pop();            //出队一个结点p
                cursum+=(double)(p->val);			//当前层的结点值累计到cursum
                if (p->left!=NULL)                      //有左孩子时将其进队
                    qu.push(p->left);
                if (p->right!=NULL)                     //有右孩子时将其进队
                    qu.push(p->right);
            }
            ans.push_back(cursum/n);
        }
        return ans;
    }
};

